singleheader/CMakeLists.txt

```diff
@@ -83,11 +83,18 @@ else(BASH)
 
 endif(BASH)
 
+#
+# Do not depend on singleheader files directly: depend on this target instead.
+# Otherwise the custom command may get triggered multiple times and race with itself!
+#
+add_custom_target(singleheader-files DEPENDS ${SINGLEHEADER_FILES})
+
 #
 # Include this if you intend to #include "simdjson.cpp" in your own .cpp files.
 #
 add_library(simdjson-singleheader-include-source INTERFACE)
 target_include_directories(simdjson-singleheader-include-source INTERFACE $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>)
+add_dependencies(simdjson-singleheader-include-source singleheader-files)
 
```

Comment:
I would have added 

```
add_dependencies(simdjson-singleheader-source singleheader-files)
```

Because it feels to me that simdjson-singleheader-source *also* depends on the header files, so you have not entirely resolved the race condition. I think we want *any* target that depends on the single header files to have a dependency on the singleheader-files target.  But maybe my mental model is wrong.

Reply:
> I would have added
> 
> ```
> add_dependencies(simdjson-singleheader-source singleheader-files)
> ```
> 
> Because it feels to me that simdjson-singleheader-source _also_ depends on the header files, so you have not entirely resolved the race condition. I think we want _any_ target that depends on the single header files to have a dependency on the singleheader-files target. But maybe my mental model is wrong.

@lemire simdjson-singleheader-source depends on simdjson-singleheader-include-source, and dependencies are transitive. It might be clearer to add the redundant line though, I admit!

Reply:
@jkeiser Hmmm... Well, simdjson-singleheader depended on simdjson-singleheader-source  which depended on  simdjson-singleheader-source  and we still got a race condition.

My mental mode is like that...

Suppose that A depends on the generated files...
Suppose that B depends on A *and* on the generated files...

The CMake can look at A, start generating the files, then look at B and say... "oh, I also need to generate the files *and* wait for A to be built".

And see... that's what I think CMake can deduce from the code...

```
target_sources(simdjson-singleheader-source INTERFACE $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/simdjson.cpp>)
```

Reply:
If my mental model is wrong, please help me debug it.

Reply:
No, I'm pretty sure the race condition is limited to add_custom_command, which works more like a bash alias than a make target. If A depends on B and B depends on C, then B will not start until C is done, and C will not start until B is done.

Reply:
@lemire with the add_custom_target, here is my mental model:

* amalgamate_demo requires ...
  * simdjson-singleheader-include-source, which requires ...
    * singleheader-files, which requires ...
      * the generated files, which are part of a *command* rather than a target and triggers immediately and with no other considerations.
* simdjson-singleheader requires ...
  * simdjson-singleheader-source, which requires ...
    * simdjson-singleheader-include-source, which requires ...
      * singleheader-files, which is also transitively required by amalgamate_demo, so it just waits on that.

Without the add_custom_target, here is what seems like happens:

* amalgamate_demo requires ...
  * simdjson-singleheader-include-source, which requires ...
    * the generated files, which are part of a *command* rather than a target and triggers immediately and with no * simdjson-singleheader requires ...
  * simdjson-singleheader-source, which requires ...
    * simdjson-singleheader-include-source, which requires ...
      * the generated files, which are part of a *command* rather than a target and triggers immediately and with no

Reply:
@jkeiser  Ok. I took some time to elaborate on my point....

https://github.com/lemire/data_races_in_cmake

Can you explain where I am going wrong?

Reply:
Basically, in my example, I simulate your scenario with 
```
target_link_libraries(simdjson-singleheader-source INTERFACE simdjson-singleheader-include-source)
```

And if you build in verbose mode, with threads, you see that it starts building both of them at the same time.

Reply:
So if I understand my own example correctly, you need to state explicitly the dependencies. It is not enough to have `target_link_libraries(A INTERFACE B)` to conclude that A won't start building before B is ready.

Reply:
@lemire the issue there is you are building non-INTERFACE libraries (libraries with actual source) but you specify the dependencies as INTERFACE. If you say INTERFACE, it means "don't apply this dependency to ME, just apply it to my dependencies." If you say PUBLIC, it means "apply this both to me and my dependencies." Changing both lines to public fixes the race:

```cmake
add_library(joe1 STATIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/textfile.cpp> )
add_library(joe2 STATIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/textfile.cpp> )
add_library(joe3 STATIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/textfile.cpp> )
target_link_libraries(joe2 PUBLIC joe1)
target_link_libraries(joe3 PUBLIC joe2)

add_custom_target(fake DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/textfile.cpp)
add_dependencies(joe1 fake)
```

Reply:
@lemire this doesn't repro the original problem, though, which blows MY theory out of the water too.

```cmake
add_custom_command(
    OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/textfile.cpp
    COMMAND ${CMAKE_COMMAND} -E env echo "allo" && pwd && ${CMAKE_CURRENT_SOURCE_DIR}/script.sh && echo "file outputted"
)
add_library(joe1 STATIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/textfile.cpp> )
add_library(joe2 STATIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/textfile.cpp> )
add_library(joe3 STATIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/textfile.cpp> )
```

Reply:
Ok. I think I get it. This is a bit crazy stuff. This interface stuff is crazy. "I don't have any dependency, but if you depend on me, then you have this dependency". 
Ok.

Reply:
@lemire never mind. My theory is sound. My last comment does in fact repro the issue.

Reply:
@lemire here is the fix. A custom target is the only thing that does it. I can only conclude that custom commands ALWAYS have to be filtered through custom targets.

```c++
add_library(joe1 STATIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/textfile.cpp> )
add_library(joe2 STATIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/textfile.cpp> )
add_library(joe3 STATIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/textfile.cpp> )

# These work:
add_custom_target(includetarget DEPENDS $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/textfile.cpp>)
add_dependencies(joe1 includetarget)
add_dependencies(joe2 includetarget)
add_dependencies(joe3 includetarget)
# But this races:
# add_library(includeme INTERFACE)
# add_dependencies(includeme INTERFACE $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/textfile.cpp> )
# target_link_libraries(joe1 includeme)
# target_link_libraries(joe2 includeme)
# target_link_libraries(joe3 includeme)

Reply:
(The issue I had was at some point I stopped running parallel make, so NOTHING ever reproed :)

Reply:
Thanks for checking things out with me.

Reply:
I'm guessing the reason for the above is actually that INTERFACE libraries don't ever create a target for you to synchronize on--they are just there to feed dependencies to things that include them.

Reply:
So to conclude: custom commands are NOT targets, they are just that: commands that trigger when you ask for the files in them. They may run multiple times, and are intended to do so, recalculating whatever they do each time. If it isn't thread safe, you must synchronize it using actual targets. (Which interface libraries are not.)

Reply:
> So to conclude: custom commands are NOT targets, they are just that: commands that trigger when you ask for the files in them. They may run multiple times, and are intended to do so, recalculating whatever they do each time. If it isn't thread safe, you must synchronize it using actual targets. (Which interface libraries are not.)


Pfft! I knew that!